### PR TITLE
avoid breaking users calling forbidden private api conanfile.__init__

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Fix: Avoid breaking users calling forbidden private api ``conanfile.__init__``.
Docs: Omit

Close https://github.com/conan-io/conan/issues/8741

#tags: slow

### CODE REVIEW

File: conans/client/loader.py [hunk 1]
```diff
@@ -47,8 +47,8 @@ def load_basic_module(self, conanfile_path, lock_python_requires=None, user=None
         """
         cached = self._cached_conanfile_classes.get(conanfile_path)
         if cached and cached[1] == lock_python_requires:
-            conanfile = cached[0](self._output, self._runner, display, user, channel,
-                                  self._requester)
+            conanfile = cached[0](self._output, self._runner, display, user, channel)
+            conanfile._conan_requires = self._requester
```
Comment (czoido) [inline conans/client/loader.py]:
```suggestion
            conanfile._conan_requester = self._requester
```

**Comment by SSE4** [comment]
@memsharded is it really needed to be fixed? that appears to be private API, thus breakage is totally expected. (and if breakage of private API is not expected, then this particular fix breaks behavior of 1.35.0).
I suggest to close #8741 instead of adding workarounds to the codebase.

**Review by czoido — APPROVED** [review]
I think we all agree this is not the ideal solution as commented in the PR: https://github.com/conan-io/conan/issues/8741#issuecomment-811158727 but I also would prefer to give some time for this users to migrate and revert this changes in the near future. 
Although it's already stated, maybe we could emphasise a bit more in the documentation that making use of the private implementation could lead to breakage.

**Comment by SSE4** [comment]
if we want to avoid breakage, this cannot be merged as is, because right now we have two categories of users:
1. pre-1.35 users who used private ctor with 4 arguments
2. 1.35 users who used private ctor with 5 arguments

the changes breaks second category now, as there is no 5 arguments ctor any more. IMO it would be better to provide two overloads of ctor then.

**Comment by czoido** [comment]
> 1. pre-1.35 users who used private ctor with 4 arguments
> 2. 1.35 users who used private ctor with 5 arguments
> 
> the changes breaks second category now, as there is no 5 arguments ctor any more. IMO it would be better to provide two overloads of ctor then.

Technically, yes, we will break people who used that 5 arguments constructor between the release of 1.35.0 and the moment we release 1.35.1, but I hope no one decided to do this in this short period of time... so IMHO I would try to help those users that have hundreds of packages affected by this now and fix this in the near future...

**Pull request merged**